include/Dialect/LWE/IR/LWEAttributes.td

```diff
@@ -55,10 +67,143 @@ def LWE_BitFieldEncoding: LWE_EncodingAttr<"BitFieldEncoding", "bit_field_encodi
+
+    %evals = arith.constant <[1, 2, 4, 5]> : tensor<4xi16>
+    // Note no `intt` operation exists in poly yet.
+    %poly1 = poly.intt %evals : tensor<4xi16> -> !poly.poly<#ring, #eval_encoding>
+    %poly2 = poly.intt %evals : tensor<4xi16> -> !poly.poly<#ring, #eval_encoding>
+    %rlwe_ciphertext = tensor.from_elements %poly1, %poly2 : tensor<2x!poly.poly<#ring, #eval_encoding>>
+    ```
+  }];
 }
+
+def RLWE_PolyRoundedEmbeddingEncoding
+  : LWE_EncodingAttrWithScalingFactor<"PolyRoundedEmbeddingEncoding", "poly_rounded_embedding_encoding"> {
+  let summary = "An attribute describing encoded RLWE plaintexts via the rounded inverse canonical embedding.";
+  let description = [{
+    Let $n$ be the degree of the polynomials in the plaintext space.
+    A "rounded embedding encoding" of a list of real or complex values $v_1,
+    \dots, v_{n/2}$ is (almost) the inverse of the following decoding map.
+
+    Map a polynomial $p \in \mathbb{Z}[x] / (x^N + 1) \to \mathbb{C}^{N/2}$
+    by evaluating it at the following $N/2$ points, where
+    $\omega = e^{2 \pi i / 2N}$ is the primitive $2N$th root of unity:
+
+    \[
+      \omega, \omega^3, \omega^5, \dots, \omega^{N-1}
+    \]
+
+    This map is called the "canonical embedding" $\tau$, though conjugates are
+    discarded for efficiency because $p(\overline{z}) = \overline{p(z)}$.
+
+    Then the complete decoding operation is
+    $\textup{Decode}(p) = (1/\Delta)\tau(p)$, where $\Delta$ is a scaling
+    parameter.
+
+    The encoding operation is the inverse of the decoding operation, but
+    because inputs may be any complex numbers, a rounding step is required
+    to invert $\tau$. There are multiple rounding mechanisms, and this
+    attribute does not specify which is used, because in theory two ciphertexts
+    that have used different roundings are still compatible, though they may
+    have different noise growth patterns.
+
+    // TODO: does it make sense to use the bit-field parameters here? It may
+    // not work sensibly because the rounding happens after scaling-post-INTT,
+    // though you can still use the parameters in this attribute to infer the
+    // right power of two to use.
+    The scaling parameter $\Delta$ is specified by the `cleartext_start` and
+    `cleartext_bitwidth` parameters, which are applied coefficient-wise using
+    the same semantics as the `bit_field_encoding`.
+
+    This attribute can be used in multiple ways:
+
+    - On a `poly.poly`, it asserts that the polynomial has been transformed
+      from a coefficient list using the canonical embedding.
+    - On a tensor of `poly.poly`, it asserts that the tensor is an RLWE
+      ciphertext for some RLWE scheme that supports the approximate embedding
+      encoding.
+
+    A typical flow for the CKKS scheme using this encoding would be to apply an
+    INTT operation to invert the canonical embedding to be a polynomial with
```

Comment:
INTT -> inverse FFT

Reply:
Are you saying that because it actually does a (real/complex) FFT in the classical sense? Or to make the prose more general because there are multiple possible implementations and "FFT" can be viewed as generically capturing all options (algorithmically)

Reply:
Yes, I meant the classical complex FFT. See for example the openFHE implementation: https://github.com/openfheorg/openfhe-development/blob/4ebb28ea7bdd894a73bc5b73e59fcfbc78253308/src/pke/lib/encoding/ckkspackedencoding.cpp#L265